### DISCUSSION TITLE
Does `loadPyodide` leak memory?

### DISCUSSION THREAD

**Discussion by suhdonghwi** [pyodide/pyodide#4338] [Q&A] [answered]
First of all thank you much for this great project! 💘 

I'm currently using Pyodide in my project and discovered that I need to re-initialize the Pyodide instance to completely reset the interpreter's state (#703).

So, I've been calling `loadPyodide` every time a user runs Python code. However, it seems like the memory usage keeps going up:

<img width="349" alt="image" src="https://github.com/pyodide/pyodide/assets/8275026/713c78f2-4ada-49a5-8045-eedd617287fc">

About ~5,000K live objects are newly created each time, and they are not getting destroyed later on.

Is `loadPyodide` causing a memory leak, or am I missing something? Thanks a lot!

**Comment by hoodmane** [comment, marked as answer]
I believe that it does leak memory. I've investigated and it's not clear to me how or under what circumstances it's possible to get the browser to reclaim the wasm memory. It's possible to forcibly detach buffers using `structuredClone` and `transfer`, but this fails for any buffer that has been used for the memory for a wasm module. I tried walking all the data structures and setting all object fields to undefined, but I was unable to tell that this was helping. One thing that definitely works is to put each instance in a separate worker and destroy the worker when you are done.

I think I'd need to ask someone more knowledgeable about v8's garbage collector for help with this. If you want to try I can give you the "rip everything up" code and see if it works any better for you.

**Reply by suhdonghwi** [reply]
Thank you, putting each Pyodide instance in a separate worker seems like a fair workaround here. I'd also like to see what your "rip everything up" code looks like 😃
